### ISSUE TITLE
Tidb memory allocation is abnormal when running business, which may lead to OOM

### ISSUE THREAD

**Issue by Lily2025** [pingcap/tidb#29390]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
oltp_004_test
run tpcc and do some DDl

### 2. What did you expect to see? (Required)
tidb memory allocation is normal

### 3. What did you see instead (Required)
Tidb memory allocation is abnormal when running business，and sometimes causing OOM
![image](https://user-images.githubusercontent.com/84712107/140008971-1b28a73a-70e5-4555-a2d0-8fab00600005.png)

### 4. What is your TiDB version? (Required)
/ # ./tidb-server -V
Release Version: v5.2.0-nightly
Edition: Community
Git Commit Hash: 318030adefc22eb94e6c9d8633e01096188cfb6f
Git Branch: heads/refs/tags/v5.2.0-nightly
UTC Build Time: 2021-11-03 01:39:41
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

**Comment by winoros** [comment]
This one is related to https://github.com/pingcap/tidb/issues/29216.
In https://github.com/pingcap/tidb/pull/29230, we used the `approximate_keys` from PD side to get the number of keys to solve this problem.（If approximate keys are 100x larger than the count in stats_meta, we use the approximate keys to calc the sample rate).

We are using (110000/count in stats_meta) or (150000/approximate keys) as the sample rate.

But In this test, we met an unsolved scenario:
- We use BR/Lightning to restore the data.
- After restoring, we begin to test immediately.
- There're many tables and they are not small. So analyzing each one is not very fast.
- During analyzing some tables, the row_count in `stats_meta` increased.
- Though the count in stats_meta is still small, it's not 100x smaller than the approximate keys.
- We choose the stats_meta to calculate the sample rate. 
- The sample rate will be larger than expected, then we collected too many samples and the memory usage increased.

From the graph pasted, the memory is always used less than previous. Because the row_count in stats_meta is increasing. Finally, the sample rate is calculated more and more closer to the expected one. And we are going to use less memory which meets our expectations.
